docs/DOCUMENTATION.md

```diff
@@ -0,0 +1,872 @@
+  - [Intervals](#consid4)
+## Oracle Network Structure
+
+<img src="images/struct.png"></img>
+<br>
+
+Each node operator has to run:
+- A Chainlink node hosted locally in Docker that queries Chainlink adapters
+- Chainlink adapters hosted beside the Chainlink node that act as middleware between the node as various 3rd party APIs
+- Middleware which initiates jobs on the Chainlink node through its API if a round hasn't been created in 1 minute on the on chain aggregator contract. Rounds are also created if the oracle operator's current price is deviated past X% from the latest on-chain median price. Oracle operators are not allowed to create 2 rounds subsequently to prevent spamming.
+- An Agoric node hosted on their own infrastructure to allow the middleware / oracle proxy to communicate and broadcast transactions reliably to the network.
+
+
+## Smart Contract Details
+
+Smart Contract - https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/price/priceAggregatorChainlink.js
+
+- The smart contract resembles a native Chainlink Solidity Flux-Monitor smart contract on Ethereum.
+- For each feed, there will be a smart contract on chain and each node operator receives invitations to their smart wallet to be part of the oracle set. Then, oracles can use that invitation to push prices on chain.
+- There is a minimum number of submissions which has to be reached in order for the on-chain price to be updated
+- An oracle cannot initialise multiple consecutive rounds to avoid network congestion
+- The updated price is the median from the submitted prices for a particular round
+- Submissions for old rounds cannot be submitted and will be rejected
+
+## Aims/Objectives
+
+#### Middleware
+
+The middleware is needed to:
+1. Create CL jobs. Chainlink Jobs are created every minute. A cron job is not used in order to be able to pass in particular inputs to the job run so that the progress of the request can be stored and monitored.
+2. Get results back from the CL node.
+3. Query smart contract for new rounds.
+3. Push price updates on chain.
+
+The middleware should contain the following functionalities:
+
+1. A state file which is updated so that whenever the middleware is restarted it is able to continue from where it was stopped.
+2. An endpoint to listen for new and removed jobs from the CL node. This is needed in order to keep a list of jobs for which the middleware has to send requests.
+3. Keep a list of these jobs in its state
+4. Query the price on chain and the latest round every X seconds so that a new job request is created if a new round is found
+5. Send CL job requests to the CL node with parameters including a request ID and the reason for the request every Y seconds. The reason can be one of the following 3 reasons:
+  a. Type 1: Time expired and a new request have to be made
+  b. Type 2: A price deviation from the on-chain price was found
+  c. Type 3: A new round was found
```

Comment:
Why is Type 1 required? a new round opens every 1 minute of no prices updates right? so why would there be a time expired request?

Reply:
Not why don't you query the smart contracts and if a query to CL node is required you issue it directly (sortof event-based) instead of having 2 cron jobs